src/parser/daphnedsl/DaphneDSLVisitor.cpp

```diff
@@ -643,52 +644,137 @@ antlrcpp::Any DaphneDSLVisitor::visitParanthesesExpr(DaphneDSLGrammarParser::Par
+    }
+
+    // UDF with the provided name does not exist
+    return std::nullopt;
+}
+
+
+std::optional<mlir::FuncOp> DaphneDSLVisitor::findMatchingUnaryUDF(const std::string &functionName, mlir::Type argType) const {
+    // search user defined functions
+    auto range = functionsSymbolMap.equal_range(functionName);
+    
+    // TODO: find not only a matching version, but the `most` specialized    
+    for (auto it = range.first; it != range.second; ++it) {
+        auto userDefinedFunc = it->second;
+        auto funcTy = userDefinedFunc.getType();
+
+        if (funcTy.getNumInputs() != 1) {
+            continue;
+        }
+
+        if (argAndUDFParamCompatible(argType, funcTy.getInput(0))) {
+            return userDefinedFunc;
         }
     }
+    // UDF with the provided name exists, but no version matches the argument types
     if (range.second != range.first) {
         // FIXME: disallow user-defined function with same name as builtins, otherwise this would be wrong behaviour
-        throw std::runtime_error("No function definition of `" + func + "` found with matching types");
+        throw std::runtime_error("No function definition of `" + functionName + "` found with matching types");
+    }
+
+    // UDF with the provided name does not exist
+    return std::nullopt;
+}
+
+antlrcpp::Any DaphneDSLVisitor::handleMapOpCall(DaphneDSLGrammarParser::CallExprContext * ctx) {
+    std::string func = ctx->func->getText();
+    assert(func == "map" && "Called 'handleMapOpCall' for another function");
+    mlir::Location loc = utils.getLoc(ctx->start);
+    
+    if (ctx->expr().size() != 2) {
+      throw std::runtime_error(
+                "built-in function 'map' expects exactly 2 argument(s), but got " +
+                std::to_string(ctx->expr().size())
+        );
+    }
+
+    std::vector<mlir::Value> args;
+    
+    auto argVal = utils.valueOrError(visit(ctx->expr(0)));
+    args.push_back(argVal);
+
+    auto argMatTy = argVal.getType().dyn_cast<mlir::daphne::MatrixType>();
+    if (!argMatTy)
+        throw std::runtime_error("built-in function 'map' expects argument of type matrix as its first parameter");
+    
+    std::string udfName = ctx->expr(1)->getText();
+    auto maybeUDF = findMatchingUnaryUDF(udfName, argMatTy.getElementType());
+
+    if (!maybeUDF)
+        throw std::runtime_error("No function definition of `" + udfName + "` found");
+
+    args.push_back(static_cast<mlir::Value>(
+        builder.create<mlir::daphne::ConstantOp>(loc, maybeUDF->sym_name().str())
+    ));
+
+    // Create DaphneIR operation for the built-in function.
+    return builtins.build(loc, func, args);
+}
+
+
+antlrcpp::Any DaphneDSLVisitor::visitCallExpr(DaphneDSLGrammarParser::CallExprContext * ctx) {
+    std::string func = ctx->func->getText();
+    mlir::Location loc = utils.getLoc(ctx->start);
+ 
+    if (func == "map") 
```

Comment:
As discussed, we need to give function `map` an extra treatment here since the second parameter is a UDF identifier and `visitIdentifierExpr` only searches the `symbolTable` which does not include UDFs.

Reply:
The current implementation is totally fine :) .